File: shopGunSdk/src/main/java/com/shopgun/android/sdk/pagedpublicationkit/PagedPublicationFragment.java
package com.shopgun.android.sdk.pagedpublicationkit;

import android.content.res.Configuration;
import android.graphics.PorterDuff;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import android.support.v4.view.CenteredViewPager;
import android.support.v4.view.ViewPager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.shopgun.android.materialcolorcreator.MaterialColor;
import com.shopgun.android.materialcolorcreator.MaterialColorImpl;
import com.shopgun.android.sdk.R;
import com.shopgun.android.sdk.utils.SgnUtils;
import com.shopgun.android.verso.VersoFragment;
import com.shopgun.android.verso.VersoPageView;
import com.shopgun.android.verso.VersoPageViewFragment;
import com.shopgun.android.verso.VersoTapInfo;
import com.shopgun.android.verso.VersoViewPager;
import com.shopgun.android.verso.VersoZoomPanInfo;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class PagedPublicationFragment extends VersoFragment {

    public static final String TAG = PagedPublicationFragment.class.getSimpleName();

    public static final String ARG_CONFIGURATION = "arg_config";
    public static final String ARG_PAGE = "arg_page";
    public static final String SAVED_STATE = "saved_state";

    PagedPublicationConfiguration mConfig;

    FrameLayout mFrame;
    FrameLayout mFrameVerso;
    FrameLayout mFrameLoader;
    FrameLayout mFrameError;
    VersoViewPager mVersoViewPager;

    boolean mDisplayHotspotsOnTouch = true;
    OnTouchWrapper mOnTouchWrapper;
    PageChangeListener mPageChangeLisetner = new PageChangeListener();

    PagedPublicationLifecycle mLifecycle;
    String mViewSessionUuid;

    public static PagedPublicationFragment newInstance() {
        return new PagedPublicationFragment();
    }

    public static PagedPublicationFragment newInstance(PagedPublicationConfiguration config) {
        return newInstance(config, 0);
    }

    public static PagedPublicationFragment newInstance(PagedPublicationConfiguration config, int page) {
        Bundle args = new Bundle();
        args.putParcelable(ARG_CONFIGURATION, config);
        args.putInt(ARG_PAGE, page);
        PagedPublicationFragment f = newInstance();
        f.setArguments(args);
        return f;
    }

    public PagedPublicationFragment() {
        mOnTouchWrapper = new OnTouchWrapper();
        super.setOnTapListener(mOnTouchWrapper);
        super.setOnDoubleTapListener(mOnTouchWrapper);
        super.setOnLongTapListener(mOnTouchWrapper);
        super.setOnZoomListener(mOnTouchWrapper);
        super.addOnPageChangeListener(mPageChangeLisetner);
        super.setOnLoadCompleteListener(mOnTouchWrapper);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mLifecycle = new PagedPublicationLifecycle();
        mViewSessionUuid = SgnUtils.createUUID();
        if (savedInstanceState != null && savedInstanceState.containsKey(SAVED_STATE)) {
            SavedState savedState = savedInstanceState.getParcelable(SAVED_STATE);
            if (savedState != null) {
                mConfig = savedState.config;
                mDisplayHotspotsOnTouch = savedState.displayHotspotOnTouch;
                mViewSessionUuid = savedState.viewSessionUuid;
                setupConfigAndTrackers();
            }
        } else if (getArguments() != null) {
            mConfig = getArguments().getParcelable(ARG_CONFIGURATION);
            if (getArguments().containsKey(ARG_PAGE)) {
                int page = getArguments().getInt(ARG_PAGE);
                setPage(page);
            }
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        mVersoViewPager = (VersoViewPager) super.onCreateView(inflater, container, savedInstanceState);
        mFrame = (FrameLayout) inflater.inflate(R.layout.shopgun_sdk_pagedpublication, container, false);
        mFrameVerso = (FrameLayout) mFrame.findViewById(R.id.verso);
        mFrameError = (FrameLayout) mFrame.findViewById(R.id.error);
        mFrameLoader = (FrameLayout) mFrame.findViewById(R.id.loader);
        setVisible(false, false, false);

        if (getVersoSpreadConfiguration() == null) {
            setPublicationConfiguration(mConfig);
        }

        return mFrame;
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        // Resume lifecycle now, so callbacks for pagechanges is registered correctly when state is restored
        setupConfigAndTrackers();
        super.onViewStateRestored(savedInstanceState);
    }

    private void setupConfigAndTrackers() {
        Configuration config = getResources().getConfiguration();
        mConfig.onConfigurationChanged(config);
        mLifecycle.setConfig(mConfig);
        mLifecycle.resetSpreadsPagesLoadedAndZoom();
    }

    public PagedPublicationConfiguration getPublicationConfiguration() {
        return mConfig;
    }

    public void setPublicationConfiguration(PagedPublicationConfiguration configuration) {
        if (configuration != null) {
            mConfig = configuration;
            mLifecycle.setConfig(mConfig);
            setVersoSpreadConfiguration(mConfig);
            loadPagedPublication();
        }
    }

    @Override
    protected void onInternalPause() {
        if (isCurrentSpreadScaled()) {
            resetCurrentSpreadScale(false);
        }
        mLifecycle.spreadDisappeared(getPosition(), getCurrentPages());
        mLifecycle.resetSpreadsPagesLoadedAndZoom();
        super.onInternalPause();
    }

    @Override
    protected void onInternalResume(Configuration config) {
        mLifecycle.resumed();
        super.onInternalResume(config);
    }

    @Override
    public void onResume() {
        super.onResume();
        loadPagedPublication();
    }

    @Override
    public void onPause() {
        super.onPause();
        mLifecycle.paused();
        if (mConfig != null) {
            mConfig.cancel();
        }
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        if (mConfig != null) {
            mConfig.cancel();
        }
        outState.putParcelable(SAVED_STATE, new SavedState(this));
        super.onSaveInstanceState(outState);
    }

    private void loadPagedPublication() {
        if (mConfig != null) {
            if (mConfig.isLoading()) {
                // loader is already working on it,
                return;
            }
            if (mConfig.getPublication() != null &&
                            mConfig.hasPages() &&
                            mConfig.hasHotspotCollection()) {
                // Already have the needed data
                notifyVersoConfigurationChanged();
                return;
            }
            showLoaderView();
            mConfig.load(new PagedPublicationOnLoadComplete());
        }
    }

    private void ensurePublicationBranding() {
        PagedPublication pub = mConfig == null ? null : mConfig.getPublication();
        if (pub != null && mFrame != null) {
            int bgColor = pub.getBackgroundColor();
            mFrame.setBackgroundColor(bgColor);
            ProgressBar bar = (ProgressBar) mFrameLoader.findViewById(R.id.circulareProgressBar);
            if (bar != null && bar.getIndeterminateDrawable() != null) {
                bar.setIndeterminate(true);
                MaterialColor c = new MaterialColorImpl(bgColor);
                bar.getIndeterminateDrawable().setColorFilter(c.getSecondaryText(), PorterDuff.Mode.MULTIPLY);
            }
        }
    }

    private void showVersoView() {
        if (mFrameVerso != null &&
                mFrameVerso.getVisibility() != View.VISIBLE) {
            ensurePublicationBranding();
            mFrameVerso.removeAllViews();
            mFrameVerso.addView(mVersoViewPager);
            mVersoViewPager.addOnPageChangeListener(new PageScrolledListener());
            setVisible(true, false, false);
        }
        mLifecycle.appeared();
    }

    private void showLoaderView() {
        if (mFrameLoader != null &&
                mFrameLoader.getVisibility() != View.VISIBLE) {
            ensurePublicationBranding();
            setVisible(false, true, false);
        }
    }

    private void showErrorView(PublicationException ex) {
        if (mFrameError != null &&
                mFrameError.getVisibility() != View.VISIBLE) {
            ensurePublicationBranding();
            mFrameError.removeAllViews();
            View errorView = getErrorView(mFrame, ex);
            mFrameError.addView(errorView);
            setVisible(false, false, true);
        }
    }

    private void setVisible(boolean verso, boolean loader, boolean error) {
        if (mFrame != null) {
            // TODO: 10/11/16 Add animation between the views
            mFrameVerso.setVisibility(verso ? View.VISIBLE : View.GONE);
            mFrameLoader.setVisibility(loader ? View.VISIBLE : View.GONE);
            mFrameError.setVisibility(error ? View.VISIBLE : View.GONE);
        }
    }

    public View getErrorView(ViewGroup container, PublicationException ex) {
        LayoutInflater i = LayoutInflater.from(container.getContext());
        View v = i.inflate(R.layout.shopgun_sdk_pagedpublication_error, container, false);
        TextView msg = (TextView) v.findViewById(R.id.message);
        msg.setText(ex.getMessage());
        return v;
    }

    private class PagedPublicationOnLoadComplete implements PagedPublicationConfiguration.OnLoadComplete {
        @Override
        public void onPublicationLoaded(PagedPublication publication) {
            notifyVersoConfigurationChanged();
        }

        @Override
        public void onPagesLoaded(List<? extends PagedPublicationPage> pages) {
            notifyVersoConfigurationChanged();
        }

        @Override
        public void onHotspotsLoaded(PagedPublicationHotspotCollection hotspots) {
            notifyVersoConfigurationChanged();
        }

        @Override
        public void onError(List<PublicationException> ex) {
            if (!mConfig.hasData()) {
                showErrorView(ex != null && !ex.isEmpty() ?
                        ex.get(0) : new PublicationException("Unknown error occurred"));
            }
        }
    }

    @Override
    public void notifyVersoConfigurationChanged() {
        if (mConfig.hasData()) {
            mLifecycle.opened();
            showVersoView();
            super.notifyVersoConfigurationChanged();
        }
    }

    public boolean isDisplayHotspotsOnTouch() {
        return mDisplayHotspotsOnTouch;
    }

    public void setDisplayHotspotsOnTouch(boolean displayHotspotsOnTouch) {
        mDisplayHotspotsOnTouch = displayHotspotsOnTouch;
    }

    @Override
    public void setOnTapListener(VersoPageViewFragment.OnTapListener tapListener) {
        mOnTouchWrapper.mTapListener = tapListener;
    }

    @Override
    public void setOnDoubleTapListener(VersoPageViewFragment.OnDoubleTapListener doubleTapListener) {
        mOnTouchWrapper.mDoubleTapListener = doubleTapListener;
    }

    public void setOnHotspotTapListener(OnHotspotTapListener tapListener) {
        mOnTouchWrapper.mHotspotTapListener = tapListener;
    }

    @Override
    public void setOnLongTapListener(VersoPageViewFragment.OnLongTapListener longTapListener) {
        mOnTouchWrapper.mLongTapListener = longTapListener;
    }

    public void setOnHotspotLongTapListener(OnHotspotLongTapListener longTapListener) {
        mOnTouchWrapper.mHotspotLongTapListener = longTapListener;
    }

    @Override
    public void setOnZoomListener(VersoPageViewFragment.OnZoomListener zoomListener) {
        mOnTouchWrapper.mZoomListener = zoomListener;
    }

    public interface OnHotspotTapListener {
        void onHotspotsTap(List<PagedPublicationHotspot> hotspots);
    }

    public interface OnHotspotLongTapListener {
        void onHotspotsLongTap(List<PagedPublicationHotspot> hotspots);
    }

    private class OnTouchWrapper implements VersoPageViewFragment.OnTapListener,
            VersoPageViewFragment.OnLongTapListener,
            VersoPageViewFragment.OnDoubleTapListener,
            VersoPageViewFragment.OnZoomListener,
            VersoPageViewFragment.OnLoadCompleteListener,
            Handler.Callback {

        private static final int DELAY = 100;
        private static final int WHAT_TAP = 1;
        private static final int WHAT_LONG_TAP = 2;

        OnHotspotTapListener mHotspotTapListener;
        VersoPageViewFragment.OnTapListener mTapListener;
        VersoPageViewFragment.OnDoubleTapListener mDoubleTapListener;
        OnHotspotLongTapListener mHotspotLongTapListener;
        VersoPageViewFragment.OnLongTapListener mLongTapListener;
        VersoPageViewFragment.OnZoomListener mZoomListener;
        VersoPageViewFragment.OnLoadCompleteListener mLoadCompleteListener;

        Handler mHandler = new Handler(Looper.getMainLooper(), this);

        @Override
        public boolean onTap(VersoTapInfo info) {
            if (mTapListener != null) {
                mTapListener.onTap(info);
            }
            PublicationTapInfo pti = new PublicationTapInfo(info);
            if (pti.hasHotspots()) {
                post(WHAT_TAP, pti);
                showHotspots(pti);
            }
            return true;
        }

        @Override
        public boolean onDoubleTap(VersoTapInfo info) {
            return mDoubleTapListener != null && mDoubleTapListener.onDoubleTap(info);
        }

        @Override
        public void onLongTap(VersoTapInfo info) {
            if (mLongTapListener != null) {
                mLongTapListener.onLongTap(info);
            }
            PublicationTapInfo pti = new PublicationTapInfo(info);
            if (pti.hasHotspots()) {
                post(WHAT_LONG_TAP, pti);
                showHotspots(pti);
            }
        }

        boolean zoomBegin = false;
        @Override
        public void onZoom(VersoZoomPanInfo info) {
            if (mZoomListener != null) {
                mZoomListener.onZoom(info);
            }
            if (zoomBegin && info.getScale() > 1.0f) {
                zoomBegin = false;
                mLifecycle.spreadZoomedIn(info);
            }
        }

        @Override
        public void onZoomBegin(VersoZoomPanInfo info) {
            mVersoViewPager.setPagingEnabled(false);
            if (mZoomListener != null) {
                mZoomListener.onZoomBegin(info);
            }
            zoomBegin = true;
        }

        @Override
        public void onZoomEnd(VersoZoomPanInfo info) {
            mVersoViewPager.setPagingEnabled(!isCurrentSpreadScaled());
            if (mZoomListener != null) {
                mZoomListener.onZoomEnd(info);
            }
            zoomBegin = false;
            mLifecycle.spreadZoomedOut(info);
        }

        @Override
        public void onPageLoadComplete(boolean success, VersoPageView versoPageView) {
            if (mLoadCompleteListener != null) {
                mLoadCompleteListener.onPageLoadComplete(success, versoPageView);
            }
            if (success) {
                mLifecycle.pageLoaded(versoPageView.getPage());
            }
        }

        private void showHotspots(PublicationTapInfo info) {
            if (!mDisplayHotspotsOnTouch) {
                return;
            }
            PagedPublicationOverlay o = getSpreadOverlay(info);
            if (o != null) {
                o.showHotspots(info);
            }
        }

        private PagedPublicationOverlay getSpreadOverlay(PublicationTapInfo info) {
            if (info != null && info.getFragment() != null) {
                View v = info.getFragment().getSpreadOverlay();
                if (v instanceof PagedPublicationOverlay) {
                    return (PagedPublicationOverlay) v;
                }
            }
            return null;
        }

        private void post(int what, PublicationTapInfo info) {
            Message msg = Message.obtain(mHandler, what, info);
            mHandler.sendMessageDelayed(msg, DELAY);
        }

        @Override
        public boolean handleMessage(Message msg) {
            switch (msg.what) {
                case WHAT_TAP:
                    if (mHotspotTapListener != null) {
                        PublicationTapInfo info = (PublicationTapInfo) msg.obj;
                        mHotspotTapListener.onHotspotsTap(info.getHotspots());
                    }
                    break;
                case WHAT_LONG_TAP:
                    if (mHotspotLongTapListener != null) {
                        PublicationTapInfo info = (PublicationTapInfo) msg.obj;
                        mHotspotLongTapListener.onHotspotsLongTap(info.getHotspots());
                    }
                    break;
            }
            return false;
        }

    }

    public class PublicationTapInfo extends VersoTapInfo {

        private final List<PagedPublicationHotspot> mHotspots;

        public PublicationTapInfo(VersoTapInfo info) {
            super(info);
            this.mHotspots = findHotspots(info);
        }

        public PublicationTapInfo(PublicationTapInfo info) {
            super(info);
            this.mHotspots = info.mHotspots;
        }

        public List<PagedPublicationHotspot> getHotspots() {
            return mHotspots;
        }

        public boolean hasHotspots() {
            return mHotspots != null && !mHotspots.isEmpty();
        }

        private List<PagedPublicationHotspot> findHotspots(VersoTapInfo info) {
            PagedPublicationHotspotCollection collection = mConfig.getHotspotCollection();
            if (collection != null && info.isContentClicked()) {
                int[] pages = Arrays.copyOf(info.getPages(), info.getPages().length);
                int introOffset = mConfig.hasIntro() ? -1 : 0;
                for (int i = 0; i < pages.length; i++) {
                    pages[i] = pages[i] + introOffset;
                }
                int pageTapped = info.getPageTapped() + introOffset;
                return collection.getPagedPublicationHotspots(pages, pageTapped, info.getPercentX(), info.getPercentY());
            }
            return new ArrayList<>();
        }

    }

    public void onPublicationDisappeared() {
        mLifecycle.saveState();
        mLifecycle.disappeared();
    }

    public void onPublicationAppeared() {
        int spread = getPosition();
        int[] pages = mConfig.getPagesFromSpreadPosition(spread);
        mLifecycle.applyState(spread, pages);
        mLifecycle.appeared();
        mLifecycle.spreadAppeared(spread, pages, true);
    }

    private class PageScrolledListener implements CenteredViewPager.OnPageChangeListener {

        int mLastState = ViewPager.SCROLL_STATE_IDLE;
        int mDragFromSpread;
        int[] mDragFromPages;

        @Override
        public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(int position) {
        }

        @Override
        public void onPageScrollStateChanged(int state) {
//            L.d(TAG, "onPageScrollStateChanged: " + ToStringUtils.pageScrollState(state));
            switch (state) {
                case ViewPager.SCROLL_STATE_DRAGGING:
                    mDragFromSpread = getPosition();
                    mDragFromPages = mConfig.getPagesFromSpreadPosition(mDragFromSpread);
                    mLifecycle.saveState();
                    mLifecycle.spreadDisappeared(mDragFromSpread, mDragFromPages);
                    break;
                case ViewPager.SCROLL_STATE_IDLE:
                    if (mDragFromSpread == getPosition() && mDragFromPages != null) {
                        // User starts dragging, but doesn't swipe to the next spread - re-appear
                        mLifecycle.applyState(mDragFromSpread, mDragFromPages);
                        mLifecycle.spreadAppeared(mDragFromSpread, mDragFromPages, true);
                    }
                    break;
                case ViewPager.SCROLL_STATE_SETTLING:
                    if (mLastState != ViewPager.SCROLL_STATE_DRAGGING) {
                        mPageChangeLisetner.mDisappearPreviousSpread = true;
                    }

            }
            mLastState = state;
        }

    }

    private class PageChangeListener implements VersoFragment.OnPageChangeListener {

        boolean mDisappearPreviousSpread = false;

        @Override
        public void onPagesScrolled(int currentPosition, int[] currentPages, int previousPosition, int[] previousPages) {

        }

        @Override
        public void onPagesChanged(int currentPosition, int[] currentPages, int previousPosition, int[] previousPages) {
            if (mDisappearPreviousSpread) {
                mLifecycle.spreadDisappeared(previousPosition, previousPages);
            }
            mDisappearPreviousSpread = false;
            mLifecycle.spreadAppeared(currentPosition, currentPages, true);
        }

        @Override
        public void onVisiblePageIndexesChanged(int[] pages, int[] added, int[] removed) {
        }

    }

    private static class SavedState implements Parcelable {

        PagedPublicationConfiguration config;
        boolean displayHotspotOnTouch;
        String viewSessionUuid;

        private SavedState(PagedPublicationFragment f) {
            config = f.mConfig;
            displayHotspotOnTouch = f.mDisplayHotspotsOnTouch;
            viewSessionUuid = f.mViewSessionUuid;
        }

        @Override
        public int describeContents() {
            return 0;
        }

        @Override
        public void writeToParcel(Parcel dest, int flags) {
            dest.writeParcelable(this.config, flags);
            dest.writeByte(this.displayHotspotOnTouch ? (byte) 1 : (byte) 0);
            dest.writeString(this.viewSessionUuid);
        }

        protected SavedState(Parcel in) {
            this.config = in.readParcelable(PagedPublicationConfiguration.class.getClassLoader());
            this.displayHotspotOnTouch = in.readByte() != 0;
            this.viewSessionUuid = in.readString();
        }

        public static final Creator<SavedState> CREATOR = new Creator<SavedState>() {
            @Override
            public SavedState createFromParcel(Parcel source) {
                return new SavedState(source);
            }

            @Override
            public SavedState[] newArray(int size) {
                return new SavedState[size];
            }
        };
    }

}

File: shopGunSdk/src/main/java/com/tjek/sdk/publicationviewer/paged/PagedPublicationFragment.kt
package com.tjek.sdk.publicationviewer.paged

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.fragment.app.viewModels
import com.shopgun.android.sdk.R
import com.tjek.sdk.DeviceOrientation
import com.tjek.sdk.api.Id
import com.tjek.sdk.api.models.PublicationV2
import com.tjek.sdk.api.remote.ErrorType
import com.tjek.sdk.api.remote.models.v2.BrandingV2
import com.tjek.sdk.getColorInt
import com.tjek.sdk.getDeviceOrientation
import com.tjek.sdk.publicationviewer.LoaderAndErrorScreenCallback
import com.tjek.sdk.publicationviewer.PublicationLoadingState
import com.tjek.sdk.publicationviewer.getDefaultErrorScreen
import com.tjek.sdk.publicationviewer.getDefaultLoadingScreen
import com.tjek.sdk.publicationviewer.paged.layouts.PublicationSpreadLayout
import com.tjek.sdk.publicationviewer.paged.libs.verso.*
import com.tjek.sdk.publicationviewer.paged.libs.verso.viewpager.CenteredViewPager

class PagedPublicationFragment :
    VersoFragment(),
    VersoPageViewListener.EventListener,
    VersoPageViewListener.OnLoadCompleteListener,
    CenteredViewPager.OnPageChangeListener
{

    private val viewModel: PagedPublicationViewModel by viewModels()
    private lateinit var ppConfig: PagedPublicationConfiguration
    private var hasSentOpenEvent: Boolean = false

    private lateinit var frame: FrameLayout
    private var frameVerso: FrameLayout? = null
    private var frameLoader: FrameLayout? = null
    private var frameError: FrameLayout? = null
    private lateinit var viewPager: VersoViewPager

    private var loadCompleteListener: OnLoadComplete? = null
    private var hotspotTapListener: OnHotspotTapListener? = null
    private var pageNumberChangeListener: OnPageNumberChangeListener? = null
    private var customScreenCallback: LoaderAndErrorScreenCallback? = null

    companion object {
        private const val arg_config = "arg_config"
        private const val arg_publication = "arg_publication"
        private const val arg_publication_id = "arg_publication_id"
        private const val saved_state = "saved_state"

        fun newInstance(
            publicationId: Id,
            configuration: PagedPublicationConfiguration
        ): PagedPublicationFragment {
            return createInstance(Bundle().apply { putString(arg_publication_id, publicationId) }, configuration)
        }

        fun newInstance(
            publication: PublicationV2,
            configuration: PagedPublicationConfiguration
        ): PagedPublicationFragment {
            return createInstance(Bundle().apply { putParcelable(arg_publication, publication) }, configuration)
        }

        private fun createInstance(args: Bundle, configuration: PagedPublicationConfiguration): PagedPublicationFragment {
            return PagedPublicationFragment().apply {
                arguments = args.also {
                    it.putParcelable(arg_config, configuration)
                }
            }
        }
    }

    fun setOnLoadCompleteListener(listener: OnLoadComplete) {
        loadCompleteListener = listener
    }

    fun setOnHotspotTapListener(listener: OnHotspotTapListener) {
        hotspotTapListener = listener
    }

    fun setOnPageChangeListener(listener: OnPageNumberChangeListener) {
        pageNumberChangeListener = listener
    }

    fun setCustomScreenCallback(callback: LoaderAndErrorScreenCallback) {
        customScreenCallback = callback
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState != null) {
            savedInstanceState.getParcelable<PagedPublicationSavedState>(saved_state)?.let { state ->
                ppConfig = state.config
                hasSentOpenEvent = state.hasSentOpenEvent
            }
        } else {
            arguments?.let {
                ppConfig = it.getParcelable(arg_config)!!
                setPage(ppConfig.initialPageNumber)
                val publication: PublicationV2? = it.getParcelable(arg_publication)
                if (publication != null) {
                    viewModel.loadPublication(publication)
                } else {
                    viewModel.loadPublication(it.getString(arg_publication_id, ""))
                }
            }
        }
        viewModel.loadingState.observe(this) { state ->
            when (state) {
                is PublicationLoadingState.Failed -> {
                    loadCompleteListener?.onError(state.error)
                    showError(state.error)
                }
                PublicationLoadingState.Loading -> showLoader()
                PublicationLoadingState.Successful -> {
                    setVersoSpreadConfiguration()
                    showVersoView()
                    notifyVersoConfigurationChanged()
                }
            }
        }

        // Forward loading of the data to host app if interested
        viewModel.publication.observe(this) {
            applyBranding(it.branding)
            loadCompleteListener?.onPublicationLoaded(it)
        }
        viewModel.pages.observe(this) { loadCompleteListener?.onPagesLoaded(it) }
        viewModel.hotspots.observe(this) { loadCompleteListener?.onHotspotLoaded(it) }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        viewPager = super.onCreateView(inflater, container, savedInstanceState) as VersoViewPager
        frame = inflater.inflate(R.layout.tjek_sdk_paged_publication, container, false) as FrameLayout
        frameVerso = frame.findViewById(R.id.verso) as FrameLayout?
        frameError = frame.findViewById(R.id.paged_error) as FrameLayout?
        frameLoader = frame.findViewById(R.id.paged_loader) as FrameLayout?
        setVisible(verso = false, loader = false, error = false)

        setOnEventListener(this)
        setOnLoadCompleteListener(this)
        viewPager.addOnPageChangeListener(this)

        return  frame
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putParcelable(saved_state, PagedPublicationSavedState(
            config = ppConfig,
            hasSentOpenEvent = hasSentOpenEvent
        ))
        super.onSaveInstanceState(outState)
    }

    private fun setVersoSpreadConfiguration() {
        var pageCount = viewModel.publication.value?.pageCount ?: 0
        var spreadCount = when (resources.configuration.getDeviceOrientation()) {
            DeviceOrientation.Landscape -> (pageCount / 2) + 1
            else -> pageCount
        }

        if (ppConfig.hasIntro) {
            pageCount++
            spreadCount++
        }
        if (ppConfig.hasOutro) {
            pageCount++
            spreadCount++
        }
        versoSpreadConfiguration = SpreadConfiguration(
            pageCount,
            spreadCount,
            spreadMargin = 0,
            introConfiguration = ppConfig.introConfiguration?.also { it.publication = viewModel.publication.value },
            outroConfiguration = ppConfig.outroConfiguration?.also { it.publication = viewModel.publication.value },
            pages = viewModel.pages.value,
            publicationBrandingColor = viewModel.publication.value?.branding?.colorHex.getColorInt(),
            deviceConfiguration = resources.configuration,
            showPageNumberWhileLoading = ppConfig.showPageNumberWhileLoading
        )
    }

    private fun applyBranding(branding: BrandingV2) {
        if (!ppConfig.useBrandColor) return
        val bgColor = branding.colorHex.getColorInt()
        if (::frame.isInitialized) {
            frame.setBackgroundColor(bgColor)
        }
    }

    private fun showVersoView() {
        frameVerso?.let {
            if (it.visibility != View.VISIBLE) {
                it.removeAllViews()
                it.addView(viewPager)
                // todo onPageChangeListener
                setVisible(verso = true, loader = false, error = false)
            }
        }

    }

    private fun showLoader() {
        val view =
            customScreenCallback?.showLoaderScreen(viewModel.publication.value?.branding) ?:
            getDefaultLoadingScreen(layoutInflater, viewModel.publication.value?.branding)
        frameLoader?.removeAllViews()
        frameLoader?.addView(view)
        setVisible(verso = false, loader = true, error = false)
    }

    private fun showError(error: ErrorType) {
        val view =
            customScreenCallback?.showErrorScreen(viewModel.publication.value?.branding, error) ?:
            getDefaultErrorScreen(layoutInflater, viewModel.publication.value?.branding, error)
        frameError?.removeAllViews()
        frameError?.addView(view)
        setVisible(verso = false, loader = false, error = true)
    }

    private fun setVisible(verso: Boolean, loader: Boolean, error: Boolean) {
        frameVerso?.visibility = if (verso) View.VISIBLE else View.GONE
        frameLoader?.visibility = if (loader) View.VISIBLE else View.GONE
        frameError?.visibility = if (error) View.VISIBLE else View.GONE
    }

    override fun onVersoPageViewEvent(event: VersoPageViewEvent): Boolean {
        return when (event) {
            is VersoPageViewEvent.Tap -> showHotspotAndNotifyListener(event.info, longPress = false)
            is VersoPageViewEvent.LongTap -> showHotspotAndNotifyListener(event.info, longPress = true)
            is VersoPageViewEvent.Touch -> {
                if (event.action == MotionEvent.ACTION_UP) {
                    event.info.fragment.spreadOverlay?.let { view ->
                        if (view is PublicationSpreadLayout) {
                            // the longTap event keep showing the hotspot until the finger goes up, so let's remove all views if any
                            view.removeHotspots()
                        }
                    }
                }
                true
            }
            else -> false
        }
    }

    // coming from the underlying Verso fragment
    override fun onPageLoadComplete(success: Boolean, versoPageView: VersoPageView?) {
        if (!success) return
        loadCompleteListener?.let { l ->
            versoPageView?.let { l.onPageLoad(it.page) }
        }
    }

    private fun showHotspotAndNotifyListener(info: VersoTapInfo, longPress: Boolean): Boolean {
        val hs = viewModel.findHotspot(info, ppConfig.hasIntro)

        // Notify listener
        if (hs.isNotEmpty()) {
            Handler(Looper.getMainLooper()).postDelayed(
                {
                    if (longPress) {
                        hotspotTapListener?.onHotspotLongTap(hs)
                    } else {
                        hotspotTapListener?.onHotspotTap(hs)
                    }

                }, 100)
        }

        // Show hotspot
        if (!ppConfig.displayHotspotsOnTouch) return false
        if (hs.isNotEmpty()) {
            info.fragment.spreadOverlay?.let { view ->
                if (view is PublicationSpreadLayout) {
                    view.showHotspots(hs, longPress)
                }
            }
        }
        return true
    }

    override fun onPageSelected(position: Int) {
        pageNumberChangeListener?.let {
            if (position < versoSpreadConfiguration.spreadCount - 1) {
                val currentPages = versoSpreadConfiguration.getPagesFromSpreadPosition(position)
                val tmpPages: IntArray = currentPages.copyOf(currentPages.size)
                if (!ppConfig.hasIntro) {
                    for (i in tmpPages.indices) {
                        tmpPages[i] += 1
                    }
                }
                it.onPageNumberChange(currentPages = tmpPages, totalPages = versoSpreadConfiguration.pageCount - 1)
            }
        }
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

    override fun onPageScrollStateChanged(state: Int) {}
}